warehouse/query-core/src/main/java/datawave/query/transformer/GroupingTransform.java

```diff
@@ -93,81 +163,185 @@ private Key createAttrKey(String fieldName) {
+                    lastKey = lastEntry.getKey();
+                    GroupingTransform.this.apply(lastEntry);
                 }
                 next = GroupingTransform.this.flush();
                 return next != null;
             }
             
             @Override
             public Entry<Key,Document> next() {
-                return next;
+                if (flatten) {
+                    log.trace("next has key: {} but I will use key: {}", next.getKey(), lastKey);
+                    return new AbstractMap.SimpleEntry<>(lastKey, next.getValue());
+                } else {
+                    // web server will aggregate the visibilities from the Attributes
+                    // and set the Document's visibility to that aggregation
+                    Document documentToReturn = next.getValue();
+                    // combine the column visibilities of all attributes in order to compute the overall document column visibility
+                    documentToReturn.setColumnVisibility(combine(documentToReturn.getAttributes().stream().filter(attr -> attr != null)
+                                    .map(attr -> attr.getColumnVisibility()).collect(Collectors.toSet())));
+                    
+                    return new AbstractMap.SimpleEntry<>(lastKey, documentToReturn);
+                }
             }
         };
-        
     }
     
     @Override
     public Entry<Key,Document> flush() {
+        if (flatten)
+            log.trace("tserver flush");
+        else
+            log.trace("webserver flush");
         if (documents == null) {
             documents = new LinkedList<>();
         }
-        if (!multiset.isEmpty()) {
-            if (log.isTraceEnabled()) {
-                log.trace("flush will use the multiset:" + multiset);
-            }
-            for (Collection<Attribute<?>> entry : multiset.elementSet()) {
-                ColumnVisibility vis;
+        if (!countingMap.isEmpty()) {
+            
+            log.trace("flush will use the countingMap: {}", countingMap);
+            
+            for (Collection<GroupingTypeAttribute<?>> entry : countingMap.keySet()) {
+                log.trace("from countingMap, got entry: {}", entry);
+                ColumnVisibility columnVisibility = null;
                 try {
-                    vis = toColumnVisibility(fieldVisibilities.get(entry));
+                    columnVisibility = toColumnVisibility(fieldVisibilities.get(entry));
                 } catch (Exception e) {
                     throw new IllegalStateException("Unable to merge column visibilities: " + fieldVisibilities.get(entry), e);
                 }
                 // grab a key from those saved during getListKeyCounts
                 Assert.notEmpty(keys, "no available keys for grouping results");
-                Key docKey = keys.get(0);
+                Key docKey = keys.get(keys.size() - 1);
```

Comment:
A comment as to why you selected the last one.  I presume because in the end you need the last key looked at to be the one returned for teardown/rebuilt purposes.